### ISSUE TITLE
tcp client endpoint syntax should support positional arguments

### ISSUE THREAD

**Issue by twisted-trac** [twisted/twisted#5358]
|[<img alt="glyph's avatar" src="https://avatars.githubusercontent.com/u/716529?s=50" width="50" height="50">](https://github.com/glyph)| @glyph reported|
|-|-|
|Trac ID|trac#5358|
|Type|enhancement|
|Created|2011-11-03 20:24:35Z|
|Branch|https://github.com/twisted/twisted/tree/positional-endpoint-args-5358|

I was a bit surprised to discover that `tcp:example.com:80` wasn't valid syntax, and instead I needed to do `tcp:host=example.com:port=80`.  I think that we should accept `tcp:example.com:80`.

Attachments:

* [tcp_positional_args.patch](https://raw.githubusercontent.com/twisted/twistedmatrix.com-trac-attachments/trunk/ticket/133/133e9dcf0c5f3587e416a390376c27d1021471fe/d06e153baeec9b70148442dbbc5b7391140883b1.patch) (6121 bytes) - added by lizp on 2012-03-12 21:14:00Z - 
* [tcp_positional_args_v2.patch](https://raw.githubusercontent.com/twisted/twistedmatrix.com-trac-attachments/trunk/ticket/133/133e9dcf0c5f3587e416a390376c27d1021471fe/58d287becb6119174e10b508028b47b694bd9e65.patch) (6291 bytes) - added by lizp on 2012-03-12 23:23:21Z - 
* [tcp_positional_args_v3.patch](https://raw.githubusercontent.com/twisted/twistedmatrix.com-trac-attachments/trunk/ticket/133/133e9dcf0c5f3587e416a390376c27d1021471fe/64cec588406695651f8977dfbe83b57677a572e8.patch) (6965 bytes) - added by lizp on 2012-03-13 17:57:57Z - 

<details><summary>Searchable metadata</summary>

```
trac-id__5358 5358
type__enhancement enhancement
reporter__glyph glyph
priority__normal normal
milestone__ 
branch__branches_positional_endpoint_args_5358 branches/positional-endpoint-args-5358
branch_author__lizp__habnabit lizp, habnabit
status__closed closed
resolution__fixed fixed
component__core core
keywords__easy easy
time__1320351875000000 1320351875000000
changetime__1331663852000000 1331663852000000
version__None None
owner__lizp lizp
cc__MostAwesomeDude
```
</details>

**Comment by twisted-trac** [comment]
|<img alt="lizp's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50">|lizp set owner to lizp<br>lizp set status to `assigned`|
|-|-|

**Comment by twisted-trac** [comment]
|<img alt="lizp's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50">|lizp removed owner<br>lizp set status to `new`|
|-|-|

**Comment by twisted-trac** [comment]
|[<img alt="habnabit's avatar" src="https://avatars.githubusercontent.com/u/21168?s=50" width="50" height="50">](https://github.com/habnabit)<a name="note_4"></a>|@habnabit commented|
|-|-|

(In [33697]) Branching to 'positional-endpoint-args-5358'

**Comment by twisted-trac** [comment]
|<img alt="Fahrenheit's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50">|Fahrenheit set owner to Fahrenheit<br>Fahrenheit set status to `assigned`|
|-|-|

**Comment by twisted-trac** [comment]
|[<img alt="habnabit's avatar" src="https://avatars.githubusercontent.com/u/21168?s=50" width="50" height="50">](https://github.com/habnabit)|@habnabit set owner to @habnabit<br>@habnabit set status to `new`|
|-|-|

**Comment by twisted-trac** [comment]
|[<img alt="habnabit's avatar" src="https://avatars.githubusercontent.com/u/21168?s=50" width="50" height="50">](https://github.com/habnabit)<a name="note_7"></a>|@habnabit set owner to lizp|
|-|-|

Hi!

    1. The coverage on this patch looks good, except for the KeyError on lines 876-877. I would put this in an `else:` and change the ifs done on `len(args)` into an if-elif chain. In this case, you wouldn't need the try-except at all.
    1. `if kwargs.get('host'):` should be `if 'host' in kwargs:` or a try-except KeyError to avoid the case when `host` is falsey (i.e. the empty string).

Thanks!

**Comment by twisted-trac** [comment]
|[<img alt="habnabit's avatar" src="https://avatars.githubusercontent.com/u/21168?s=50" width="50" height="50">](https://github.com/habnabit)<a name="note_8"></a>|@habnabit commented|
|-|-|

Oh, also--the documentation should be updated to reflect all four ways that `clientFromString('tcp:...')` can be called now.

**Comment by twisted-trac** [comment]
|<img alt="lizp's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50"><a name="note_9"></a>|lizp set status to `assigned`|
|-|-|

1. Putting that last ```kwargs['port'] = int(kwargs['port']``` inside an else doesn't make sense, really. If, for some reason, a person has ```'tcp:example.com:port=1234'```, then that code has to be run (with len(args) being 1.) 

Additionally, I wanted to preserve the original behavior when the port was unspecified (the KeyError would be caught here, and then an error message raised when the client was actually created without a port specified.)

If you think it would be better, I can change the code so that the error is raised as the arguments are parsed, but I think it makes more sense the way it was written originally.

2. I thought that we wouldn't want to have an empty host string, my mistake. Converted it to 
```'host' in kwargs```

3. Updated the documentation

**Comment by twisted-trac** [comment]
|<img alt="lizp's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50"><a name="note_10"></a>|lizp set owner to @habnabit<br>lizp set status to `new`|
|-|-|

1. Putting that last ```kwargs['port'] = int(kwargs['port']``` inside an else doesn't make sense, really. If, for some reason, a person has ```'tcp:example.com:port=1234'```, then that code has to be run (with len(args) being 1.) 

Additionally, I wanted to preserve the original behavior when the port was unspecified (the KeyError would be caught here, and then an error message raised when the client was actually created without a port specified.)

If you think it would be better, I can change the code so that the error is raised as the arguments are parsed, but I think it makes more sense the way it was written originally.

2. I thought that we wouldn't want to have an empty host string, my mistake. Converted it to 
```'host' in kwargs```

3. Updated the documentation

**Comment by twisted-trac** [comment]
|[<img alt="MostAwesomeDude's avatar" src="https://avatars.githubusercontent.com/u/118035?s=50" width="50" height="50">](https://github.com/MostAwesomeDude)<a name="note_12"></a>|@MostAwesomeDude set owner to lizp|
|-|-|

As a minor nit, I'd greatly appreciate it if the three different endpoint strings in the TCP endpoint test were split into three different test methods. Otherwise, I think this is pretty much ready. Please merge contingent on that change happening.

Thanks for your hard work! :3

**Comment by twisted-trac** [comment]
|[<img alt="cyli's avatar" src="https://avatars.githubusercontent.com/u/223200?s=50" width="50" height="50">](https://github.com/cyli)<a name="note_13"></a>|@cyli commented|
|-|-|

Thanks for working on this lizp, and thank you for the review MostAwesomeDude!

@lizp: For future patches, could you please include a news file, as per http://twistedmatrix.com/trac/wiki/ReviewProcess#Newsfiles ?

I'm going to add one to this and I'm going to merge.

**Comment by twisted-trac** [comment]
|[<img alt="cyli's avatar" src="https://avatars.githubusercontent.com/u/223200?s=50" width="50" height="50">](https://github.com/cyli)<a name="note_14"></a>|@cyli set status to `closed`|
|-|-|

Ok, looks like trac is having issues with its commit hooks again, so I'm going to close the ticket manually.
```
Merge patch tcp_positional_args_v3.patch: Support for shorter strings in t.i.endpoints.clientsFromString

Author: lizp
Reviewer: habnabit, MostAwesomeDude
Fixes: #5358

twisted.internet.endpoints.clientsFromStrong now supports strings in the form of tcp:example.com:80
```

**Comment by twisted-trac** [comment]
|[<img alt="cyli's avatar" src="https://avatars.githubusercontent.com/u/223200?s=50" width="50" height="50">](https://github.com/cyli)<a name="note_15"></a>|@cyli commented|
|-|-|

Addendum:  that was [revision 33737](http://twistedmatrix.com/trac/changeset/33737)
